### PR TITLE
Integration testing fetcher tooling

### PR DESCRIPTION
**Pull request by prastoin**
# Introduction

In a nutshell created easier to configure and type fetchers tools

While covering workspace/user creation and edge cases through integration tests I needed to have more granularity on fetchers services

In this way centralized all fetcher to consume the very same `supertest` client instance.
Also refactored few redundant `supertest` client initilization across some tests files ( only one or two ) as there're too many to handle right now and could be considered as legacy way to write tests ! ( remains 162 occurrences )

## Vision
In the best of the world we should never have to call atomic fetcher such as MakeGraphqlApi and Metadata fetchers but rather call already typed abstracted services such as `createOneOperation`

Migrated legacy response typing by abusively passing `any` generics, in the best of the world any new addition should be typed ( with upcoming centralized early alpha sdk stuff )

### CODE REVIEW

File: packages/twenty-server/test/integration/graphql/utils/sign-up-operation.util.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import { SignUpOutput } from 'src/engine/core-modules/auth/dto/sign-up.output';
+import { signUpOperationFactory } from 'test/integration/graphql/utils/sign-up-operation-factory.util';
+import { CommonResponseBody } from 'test/integration/types/common-response-body.type';
+import { makeMetadataAPIRequest } from 'test/integration/utils/make-metadata-api-request.util';
+
+export const performSignUp = async (
+  signUpInput: SignUpInput,
+): CommonResponseBody<{
+  findResponse: SignUpOutput;
+}> => {
```
[the remark this answers is not in the document]
Comment (prastoin) [reply packages/twenty-server/test/integration/graphql/utils/sign-up-operation.util.ts]: outdated

File: packages/twenty-server/test/integration/graphql/suites/object-records-permissions/destroy-one-object-records-permissions.integration-spec.ts [hunk 2]
```diff
@@ -16,114 +15,122 @@ describe('destroyOneObjectRecordsPermissions', () => {
...
         'IS_PERMISSIONS_V2_ENABLED',
         false,
       );
 
-      await makeGraphqlAPIRequest(disablePermissionsQuery);
+      await makeGraphqlAPIRequest({ operation });
     });
 
     it('should throw a permission error when user does not have permission (guest role)', async () => {
       const personId = randomUUID();
```
[the remark this answers is not in the document]
Comment (prastoin) [reply packages/twenty-server/test/integration/graphql/suites/object-records-permissions/destroy-one-object-records-permissions.integration-spec.ts:100]: Indeed, won't touch to existing test logic here as anw it should fail in prior due to access logic

File: packages/twenty-server/test/integration/metadata/suites/field-metadata/utils/create-one-field-metadata.util.ts [hunk 3]
```diff
@@ -2,26 +2,25 @@ import {
...
 
 export const createOneFieldMetadata = async ({
   input,
   gqlFields,
   expectToFail = false,
-}: PerformMetadataQueryParams<CreateOneFieldFactoryInput>): CommonResponseBody<{
-  createOneField: FieldMetadataEntity;
-}> => {
-  const graphqlOperation = createOneFieldMetadataQueryFactory({
+}: PerformMetadataQueryParams<CreateOneFieldFactoryInput>) => {
```
[the remark this answers is not in the document]
Comment (prastoin) [reply packages/twenty-server/test/integration/metadata/suites/field-metadata/utils/create-one-field-metadata.util.ts:15]: already infer by TypeScript

File: packages/twenty-server/test/integration/graphql/suites/object-records-permissions/restore-one-object-records-permissions.integration-spec.ts [hunk 4]
```diff
@@ -42,14 +45,19 @@ describe('restoreOneObjectRecordsPermissions', () => {
       recordId: personId,
     });
 
-    const response = await makeGraphqlAPIRequestWithGuestRole(graphqlOperation);
+    const response = await makeGraphqlAPIRequest<any>({
```
[the remark this answers is not in the document]
Comment (prastoin) [reply packages/twenty-server/test/integration/graphql/suites/object-records-permissions/restore-one-object-records-permissions.integration-spec.ts:48]: Legacy

File: packages/twenty-server/test/integration/graphql/suites/settings-permissions/workspace.integration-spec.ts [hunk 5]
```diff
@@ -60,462 +61,401 @@ describe('workspace permissions', () => {
...
-          .set('Authorization', `Bearer ${ADMIN_ACCESS_TOKEN}`)
-          .send(queryData)
-          .expect((res) => {
-            expect(res.body.data).toBeDefined();
-            expect(res.body.errors).toBeDefined();
-            expect(res.body.errors[0].message).toBe(
-              'Invalid feature flag key, flag is not public',
-            );
-          });
+        });
```
[the remark this answers is not in the document]
Comment (prastoin) [reply packages/twenty-server/test/integration/graphql/suites/settings-permissions/workspace.integration-spec.ts:425]: I think the data is defined by the field response is null.
Prefer not being too invasive in the refactor

File: packages/twenty-server/test/integration/graphql/utils/perform-create-many-operation.utils.ts [hunk 6]
```diff
@@ -1,26 +1,34 @@
...
-      objectMetadataSingularName,
-      objectMetadataPluralName,
-      gqlFields,
-      data: data.map((item) => ({
-        id: randomUUID(),
-        ...item,
-      })),
-    }),
-  );
+): CommonResponseBody<{ createManyResponse: unknown }> => {
```
[the remark this answers is not in the document]
Comment (prastoin) [reply packages/twenty-server/test/integration/graphql/utils/perform-create-many-operation.utils.ts:13]: This could be improved by the anyone needing this

File: packages/twenty-server/test/integration/utils/make-graphql-api-request.util.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import {
+  GraphqlOperationWithOptions,
+  makeGraphqlRequest,
+} from 'test/integration/utils/make-graphql-request.util';
+
+export const makeGraphqlAPIRequest = async <T>(
+  operationAndOptions: GraphqlOperationWithOptions,
+) => await makeGraphqlRequest<T>('graphql', operationAndOptions);
```
[the remark this answers is not in the document]
Comment (prastoin) [reply packages/twenty-server/test/integration/utils/make-graphql-api-request.util.ts]: Overkill

File: packages/twenty-server/test/integration/utils/make-graphql-request.util.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  API_KEY:
+    'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIyMDIwMjAyMC0xYzI1LTRkMDItYmYyNS02YWVjY2Y3ZWE0MTkiLCJ0eXBlIjoiQVBJX0tFWSIsIndvcmtzcGFjZUlkIjoiMjAyMDIwMjAtMWMyNS00ZDAyLWJmMjUtNmFlY2NmN2VhNDE5IiwiaWF0IjoxNzQ0OTgzNzUwLCJleHAiOjQ4OTg1ODM2OTMsImp0aSI6IjIwMjAyMDIwLWY0MDEtNGQ4YS1hNzMxLTY0ZDAwN2MyN2JhZCJ9.4xkkwz_uu2xzs_V8hJSaM15fGziT5zS3vq2lM48OHr0',
+} as const;
+type TestingMockAccessTokenKeys = keyof typeof TESTING_MOCK_ACCESS_TOKEN;
+
+export type MakeMetadataAPIRequestOptions = {
+  unAuthenticated?: boolean;
+  authenticationToken?: string;
+  testingToken?: TestingMockAccessTokenKeys;
+};
```
[the remark this answers is not in the document]
Comment (prastoin) [reply packages/twenty-server/test/integration/utils/make-graphql-request.util.ts]: Oh interesting

**Comment by prastoin** [comment]
@greptile trigger

**Comment by prastoin** [comment]
Thanks @martmull for the review ! as we discussed on discord the devX typing experience of the fetcher tooling isn't great.
We type things as entities but when in reality they're IConnection with pagination and edges
Closing this from now
About to do a POC where we extract the graphql codegen outside of the twenty-front in order to re-use all of the queries inputs and responses types
